db/write_batch.cc

```diff
@@ -1190,11 +1203,6 @@ class MemTableInserter : public WriteBatch::Handler {
       if (has_valid_writes_ != nullptr) {
         *has_valid_writes_ = true;
       }
-    } else {
-      // in non-recovery we ignore prepare markers
-      // and insert the values directly. making sure we have a
-      // log for each insertion to reference.
-      assert(log_number_ref_ > 0);
```

Comment:
This is incorrect in WritePrepared since we write the batch that includes the prepare maker directly to memtable, when we do not have a prepare log to ref. We moved this check to MarkCommit since in WriteCommitted, commit marker appears before the write batch when it is inserted into memtable, so it is sufficient to check there. Also in WritePrepared CommitMarker is never processed for memtable insert so MarkCommit is skipped. We have added a TODO to make the logic simpler.

Reply:
how about add a flag to write batch to let it know if it is WritePrepared or WriteCommitted? In that way the overall logic will be cleaner. Say you can still call MarkCommit in both cases but the write batch will do the work accordingly.

Reply:
Just realize with WritePrepared we don't need such logic to hold log with prepare from purge. Nice!

Reply:
My concern with acting based on an option set in WriteBatch is that the user might sometime pass the option to WriteBatch and sometime does not, and we would have to support both.